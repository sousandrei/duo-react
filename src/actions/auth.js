import { history } from '../routers/Router'
// import { get } from 'axios'

// const ENDPOINT = process.env.ENDPOINT


export const loginAction = (user) => ({
	type: 'LOGIN',
	user
})

export const login = (obj = {
	matricula: '',
	password: ''
}) => {
	return dispatch => {
		return new Promise((resolve, reject) => {

			//todo request
			let user = { ...obj, nome: 'Test' }

			//error
			// return reject('erro no login')

			//sucess
			history.push('/dashboard')
			resolve()

			dispatch(loginAction(user))

		})
	}
}

export const logoutAction = () => ({
	type: 'LOGOUT'
})

export const logout = () => {
	return dispatch => {
		dispatch(logoutAction())
		history.push('/')
	}
}
