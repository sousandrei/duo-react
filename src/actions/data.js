import axios from 'axios'

const DASHBOARD = process.env.DASHBOARD

export const getDataAction = data => ({
	type: 'GET_DATA',
	data
})

export const getData = () => {
	return async dispatch => {
		try {
			let result = await axios.get(DASHBOARD)
			dispatch(getDataAction(result.data))
		} catch (err) {
			console.error('could not retrieve data')
		}
	}
}
