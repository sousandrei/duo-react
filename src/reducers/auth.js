export default (state = {}, action) => {
	switch (action.type) {
		case 'LOGIN':
			localStorage.setItem('sirin', JSON.stringify(action.user))
			return { user: action.user }
		case 'LOGOUT':
			localStorage.removeItem('sirin')
			return {}
		default:
			return state
	}
}	