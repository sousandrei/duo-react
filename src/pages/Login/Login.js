import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { login } from '../../actions/auth'

import Loading from '../Loading/Loading'

export class Login extends React.Component {

	state = {
		error: '',
		username: '',
		senha: '',
		focused: false,
		loading: false
	}

	handleFocus = () => {
		this.setState({ focused: true })
	}

	handleBlur = () => {
		this.setState({ focused: false })
	}

	handleLogin = async e => {
		e.preventDefault()

		let senha = this.state.senha
		let username = this.state.username

		this.setState({ loading: true })

		try {
			await this.props.login({ username, senha })
		} catch (error) /* istanbul ignore next*/ {
			this.setState({ error: error.toString(), loading: false })

			// todo tratar o tipo de erro de login
			setTimeout(() => {
				this.setState({ error: '' })
			}, 3000)
		}

	}

	handleUsernameChange = e => {
		const error = 'username muito curto'
		let username = e.target.value

		username.length >= 6 ?
			this.setState(() => ({ error: '' })) :
			this.setState(() => ({ error }))

		this.setState(() => ({ username }))
	}

	handleSenhaChange = e => {
		const passError = 'Senha invalida'
		let senha = e.target.value
		let error = this.state.error

		senha.length <= 3 ?
			error ? null : error = 'Senha invalida' :
			error == passError ? error = '' : null

		senha.length < 1 ?
			error = '' : null

		this.setState(() => ({ senha, error }))
	}

	render() {
		return (
			<div className='login'>
				{this.state.loading ?
					<Loading /> :
					<form
						onSubmit={this.handleLogin}>
						<input
							onFocus={this.handleFocus}
							onBlur={this.handleBlur}
							placeholder='username'
							onChange={this.handleUsernameChange}
							value={this.state.username}
							type='text'
							name='username' />
						<input
							onFocus={this.handleFocus}
							onBlur={this.handleBlur}
							placeholder='senha'
							onChange={this.handleSenhaChange}
							value={this.state.senha}
							type='password'
							name='senha' />
						<button
							disabled={this.state.loading ||
								this.state.error.length > 1 ||
								this.state.username.length < 6 ||
								this.state.senha.length < 3}>
							<span>ENTRAR</span>
						</button>
						{this.state.error &&
							<p>{this.state.error}</p>}
					</form>
				}
			</div >
		)
	}
}

Login.propTypes = {
	login: PropTypes.func,
}


const mapDispatchToProps = dispatch => {
	return {
		login: () => dispatch(login())
	}
}

export default connect(undefined, mapDispatchToProps)(Login)
