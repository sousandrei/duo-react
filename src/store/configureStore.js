import thunk from 'redux-thunk'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'

import auth from '../reducers/auth'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

export default () => {
	const store = createStore(
		combineReducers({
			auth
		}),
		composeEnhancers(applyMiddleware(thunk))
	)

	return store
}
