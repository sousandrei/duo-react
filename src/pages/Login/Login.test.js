import React from 'react'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import configureStore from 'redux-mock-store'

import { loginAction } from '../../actions/auth'

import Login, { Login as LoginShallow } from './Login'

const mockStore = configureStore([thunk])

test('should render Login correctly', () => {
	const wrapper = shallow(<LoginShallow />)
	expect(wrapper).toMatchSnapshot()
})

test('should render Login correctly with store', () => {
	const store = mockStore()
	const wrapper = shallow(<Login store={store} />)
	expect(wrapper).toMatchSnapshot()
})

test('should call login on button click', () => {
	const login = jest.fn()
	const wrapper = shallow(<LoginShallow login={login} />)
	const matricula = 'C098098'
	const senha = '1234'

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value: matricula } })

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value: senha } })

	wrapper.find('form')
		.simulate('submit', { preventDefault: () => { } })

	expect(login).toHaveBeenCalled()
})

test('should dispatch login on button click', () => {
	const store = mockStore()
	const wrapper = mount(<Login store={store} />)

	const matricula = 'C098098'
	const senha = '1234'

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value: matricula } })

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value: senha } })

	store.subscribe(() => {
		expect(store.getActions()).toEqual([loginAction()])
	})

	wrapper.find('form')
		.simulate('submit', { preventDefault: () => { } })
})

test('should set matricula if valid input', () => {
	const value = 'C098899'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('matricula')).toBe(value)
})

test('should not set matricula if invalid input', () => {
	const value = 'ASDASDd'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeTruthy()
})

test('should not set senha if invalid input', () => {
	const value = 'asd'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeTruthy()
})

test('should handle form focus', () => {
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="senha"]')
		.simulate('focus')

	expect(wrapper.state('focused')).toBeTruthy()
})

test('should handle form blur', () => {
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="senha"]')
		.simulate('blur')

	expect(wrapper.state('focused')).toBeFalsy()
})

test('should handle matricula too big', () => {
	const value = 'C321321321asd'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('matricula')).toEqual('')
})

test('should handle remove error on matricula last digit', () => {
	let value = 'c12312c'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeTruthy()

	value = value.slice(0, value.length - 1)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeFalsy()
})

test('should handle password empty', () => {
	let value = ''
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeFalsy()

})

test('should handle password empty with matricula error', () => {
	let matricula = 'ccccccc'
	let senha = 'a'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="matricula"]')
		.simulate('change', { target: { value: matricula } })

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value: senha } })

	expect(wrapper.state('error')).toBeTruthy()

})

test('should handle password error with matricula error', () => {
	let value = 'as'
	const wrapper = shallow(<LoginShallow />)

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeTruthy()

	value = 'asasd'

	wrapper.find('input[name="senha"]')
		.simulate('change', { target: { value } })

	expect(wrapper.state('error')).toBeFalsy()

})
