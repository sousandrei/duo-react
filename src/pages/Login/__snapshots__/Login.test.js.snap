// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login correctly 1`] = `
<div
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      src="/images/marca_caixa_sirin.svg"
    />
  </div>
  <div
    className="login__form"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        name="matricula"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="matricula"
        type="text"
        value=""
      />
      <input
        name="senha"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="senha"
        type="password"
        value=""
      />
      <button
        disabled={true}
      >
        <span>
          ENTRAR
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`should render Login correctly with store 1`] = `
<Login
  login={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
